miner/Miner.hs

```diff
@@ -180,26 +185,26 @@ main = do
 
 -- | Submit a new `BlockHeader` to mine (i.e. to determine a valid `Nonce`).
 --
-submit :: Env -> BlockHeader -> IO ()
-submit (work -> w) bh = atomically $
-    isEmptyTMVar w >>= bool (void $ swapTMVar w bh) (putTMVar w bh)
+submit :: Env -> ChainId -> BlockHeight -> Bites -> IO ()
+submit (work -> w) cid bh bs = atomically $
+    isEmptyTMVar w >>= bool (void $ swapTMVar w t3) (putTMVar w t3)
+  where
+    t3 = T3 cid bh bs
 
 -- | For some `ChainId` and `BlockHeight`, have we mined a result?
 --
-poll :: Env -> ChainId -> BlockHeight -> IO (Maybe BlockHeader)
-poll (results -> tm) cid h = M.lookup (T2 cid h) <$> readTVarIO tm
-
--- | Cease mining until another `submit` call is made.
---
--- halt :: IO ()
--- halt = undefined
+poll :: Env -> ChainId -> BlockHeight -> Handler HeaderBytes
+poll (results -> tm) cid h = M.lookup (T2 cid h) <$> liftIO (readTVarIO tm) >>= \case
+    Nothing -> throwError $ err404 { errBody = "No results for given ChainId/BlockHeight" }
```

Comment:
Poll failure is now communicated via a 404 instead of a 200 + `Maybe`.

Reply:
Was that a specified change? That doesn't seem like a correct polling behavior. That's Server or Page Not Found, where a poll is simple "your request succeeded, but no corresponding data was found". It should return a 204 or 200 with empty data

Reply:
What kind of failure are we talking about?  If "poll failure" == nothing has been mined yet, then I think I agree with Emily.  That definitely shouldn't be a 404.  200 + `Maybe` would be correct.

Reply:
Let me introspect as to why I changed it in the first place (it original used `Maybe`).

Reply:
The change was because I altered the response type:
```haskell
    :<|> "poll" :> Capture "chainid" ChainId
                :> Capture "blockheight" BlockHeight
                :> Get '[OctetStream] HeaderBytes
```
It's not JSON anymore, so I can't return a `Maybe`.

Reply:
We've opted to return 200 in all cases, yielding an empty `ByteString` when lookup fails.